### ISSUE TITLE
Exceptions when traversing using sibling, adjacent and nth-child selectors and non-standard AST nodes

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#13639]
**Tell us about your environment**

* **ESLint Version:** v7.7.0
* **Node Version:** v12.14.1
* **npm Version:** v6.13.4

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

`@typescript-eslint/parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  "parser": "@typescript-eslint/parser",
  "plugins": [ "eslint-plugin-test" ],
  "rules": { "test/test": "error" }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

**`eslint-plugin-test` source:**
```js
exports.rules = {
    test: {
        create(context) {
            return {
                "Program ~ *": node => {
                    context.report({
                        message: "test",
                        node
                    })

                    // this code doesn't get a chance to get called :(
                }
            }
        }
    }
}
```

**lint file source:**
```js
const foo: string = "bar"
```

```bash
eslint ./test.ts
```

**What actually happened? Please include the actual, raw output from ESLint.**

```
Oops! Something went wrong! :(

ESLint: 7.7.0

TypeError: Cannot read property 'length' of undefined
Occurred while linting /home/samual/test/index.ts:1
    at u (/home/samual/test/node_modules/esquery/dist/esquery.min.js:1:31904)
    at Function.l [as matches] (/home/samual/test/node_modules/esquery/dist/esquery.min.js:1:30807)
    at NodeEventGenerator.applySelector (/home/samual/test/node_modules/eslint/lib/linter/node-event-generator.js:253:21)
    at NodeEventGenerator.applySelectors (/home/samual/test/node_modules/eslint/lib/linter/node-event-generator.js:281:22)
    at NodeEventGenerator.enterNode (/home/samual/test/node_modules/eslint/lib/linter/node-event-generator.js:297:14)
    at CodePathAnalyzer.enterNode (/home/samual/test/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:673:23)
    at /home/samual/test/node_modules/eslint/lib/linter/linter.js:949:32
    at Array.forEach (<anonymous>)
    at runRules (/home/samual/test/node_modules/eslint/lib/linter/linter.js:944:15)
    at Linter._verifyWithoutProcessors (/home/samual/test/node_modules/eslint/lib/linter/linter.js:1170:31)
```

**Context**

Original issue: https://github.com/typescript-eslint/typescript-eslint/issues/2439

Investigation so far:

Our parser (`@typescript-eslint/parser`) returns an extended list of [visitor-keys](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/visitor-keys) (as specified [by the ESLint docs](https://eslint.org/docs/developer-guide/working-with-custom-parsers)) [right here](https://github.com/typescript-eslint/typescript-eslint/blob/2cb66205de797479d9b2d362652c42fe032e913b/packages/parser/src/parser.ts#L140)

However `esquery` uses its own (hard coded) visitor keys for 3 selectors:
- sibling `~`:
https://github.com/estools/esquery/blob/e27e73d8cde63a2eb1aba3424510ce1b714d207e/esquery.js#L239
- adjacent `+`
https://github.com/estools/esquery/blob/e27e73d8cde63a2eb1aba3424510ce1b714d207e/esquery.js#L275
- and nth-child `:nth-child(n)`
https://github.com/estools/esquery/blob/e27e73d8cde63a2eb1aba3424510ce1b714d207e/esquery.js#L309

So I *think* that this is this something we'll have to PR into the underlying esquery library?

**Comment by mdjermanovic** [comment]
I think we should treat this as a bug, since our [selectors documentation](https://github.com/eslint/eslint/blob/master/docs/developer-guide/selectors.md) lists those selectors, so it's reasonable to expect that they can work with custom parsers.

Also, they'll crash even with the default parser on any JSX code, as in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCBcIlByb2dyYW0gfiAqXCJdICovXG5cbmNvbnN0IGZvbyA9IDxkaXYvPjsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19).

It's true that this needs to be supported in `esquery` first, so we could pass visitors keys.

Relevant PR: https://github.com/estools/esquery/pull/112

An alternative is to remove those selectors from the documentation. If this doesn't get fixed, it seems best not to use them at all.

**Comment by brettz9** [comment]
`esquery` has just published a new release with the support for custom visitor keys, so I think this should no longer be blocking.

**Comment by brettz9** [comment]
Also, `has:()` selectors are also supported which perhaps the eslint docs can indicate (the subject indicator is also supported, but per https://github.com/estools/esquery/pull/61 , that implementation is buggy while the `:has()` one was working).

**Comment by mdjermanovic** [comment]
@brettz9 can you please open a separate issue about the new esquery syntax?

ESLint selectors are using esquery, but we're also doing some optimizations and calculating specificity in [NodeEventGenerator](https://github.com/eslint/eslint/blob/master/lib/linter/node-event-generator.js), so it's possible that some additional work has to be done before we can say that ESLint fully supports new selectors.

**Comment by brettz9** [comment]
@mdjermanovic : Thanks, I've filed #14076 .
